node-packages/commons/src/tasks.ts

```diff
@@ -667,177 +648,187 @@ enum bulkType {
-  bulkName: string,
+  bulkId: string | null,
+  bulkName: string | null,
   buildPriority: number,
-  buildVariables: any,
+  buildVariables: Array<{name: string, value: string}>,
   bulkTask: bulkType
-): Promise<[string, string, string]> {
-  let projectVars: Array<Pick<EnvKeyValue, 'name' | 'value'> & {
+): Promise<{ routerPattern: string, appliedEnvVars: string }> {
+  type EnvKeyValueInternal = Pick<EnvKeyValue, 'name' | 'value'> & {
     scope: EnvVariableScope | InternalEnvVariableScope;
-  }> = [...project.envVariables];
+  }
+
+  let appliedEnvVars: EnvKeyValueInternal[] = [];
 
-    // set routerpattern to the routerpattern of what is defined in the project scope openshift
-  var routerPattern = project.openshift.routerPattern
+  // Adds newVar only if it's unique (by name).
+  const addApplied = (newVar: EnvKeyValueInternal): void => {
+    const index = appliedEnvVars.findIndex((searchVar) => searchVar.name === newVar.name)
+
+    if (index === -1) {
+      appliedEnvVars.push(newVar)
+    }
+  }
+
+  // Get the router pattern from the project or deploy target.
+  let routerPattern = project.openshift.routerPattern
+  // deployTarget.routerPattern allows null.
   if (typeof deployTarget.routerPattern !== 'undefined') {
-    // if deploytargets are being provided, then use what is defined in the deploytarget
-    // null is a valid value for routerPatterns here...
     routerPattern = deployTarget.routerPattern
   }
-  // but if the project itself has a routerpattern defined, then this should be used
+  // project.routerPattern does now allow null.
   if (project.routerPattern) {
-    // if a project has a routerpattern defined, use it. `null` is not valid here
     routerPattern = project.routerPattern
   }
 
-  projectVars = [
-    ...projectVars,
-    // append the routerpattern to the projects variables
-    // use a scope of `internal_system` which isn't available to the actual API to be added via mutations
-    // this way variables or new functionality can be passed into lagoon builds using the existing variables mechanism
-    // avoiding the needs to hardcode them into the spec to then be consumed by the build-deploy controller
-    {
-      name: 'LAGOON_SYSTEM_ROUTER_PATTERN',
-      value: routerPattern,
-      scope: InternalEnvVariableScope.INTERNAL_SYSTEM
-    },
-    // append the `LAGOON_SYSTEM_CORE_VERSION` variable as an `internal_system` variable that can be consumed by builds and
-    // is not user configurable, this value will eventually be consumed by `build-deploy-tool` to be able to reject
-    // builds that are not of a supported version of lagoon
-    {
-      name: "LAGOON_SYSTEM_CORE_VERSION",
-      value: lagoonVersion,
-      scope: InternalEnvVariableScope.INTERNAL_SYSTEM
-    }
-  ];
+  // Load organization if project is in one.
+  let org;
+  if (project.organization) {
+    org = await getOrganizationById(project.organization);
+  }
 
+  /*
+   * Internal scoped env vars.
+   */
```

Comment:
Since the `addApplied` needs to be run in a specific order to prevent variables from being overridden by others. It would be nice to make sure that this is made clear in the internal comments that the order of `addApplied` matters.

In this case, we want to ensure that a user cannot override an `internal_system` scoped variable. And then the same for when the rest of the variables are added.

Reply:
This also applies to the comments for the `normally scoped` comment, and then the following env, project, org comments.

Reply:
Renamed function and added extra comments.